internal/generator/generator_util.go

```diff
@@ -0,0 +1,200 @@
+// which specify a behavior of a directive.
+// An example is []string{"ngxHTTPMainConf", "ngxConfFlag",}.
+// A directive can have several masks.
+type mask []string
+
+type supportFileTmplStruct struct {
+	Directive2Masks map[string][]mask
+	MapVariableName string
+	MatchFnName     string
+}
+
+var (
+	// Extract single directive definition block
+	// static ngx_command_t  {name}[] = {definition}
+	// this regex extracts {name} and {definition}.
+	directivesDefineBlockExtracter = regexp.MustCompile(`ngx_command_t\s+(\w+)\[\]\s*=\s*{(.*?)}\s*;`)
+
+	// Extract one directive definition and attributes from extracted block
+	// { ngx_string({directive_name}),
+	//   {bitmask1|bitmask2|...},
+	//   ... },
+	// this regex extracts {directive_name} and {bitmask1|bitmask2|...}.
+	singleDirectiveExtracter = regexp.MustCompile(`ngx_string\("(.*?)"\).*?,(.*?),`)
+
+	singleLineCommentExtracter = regexp.MustCompile(`//.*`)
+
+	multiLineCommentExtracter = regexp.MustCompile(`/\*[\s\S]*?\*/`)
+)
+
+// Template of support file. A support file contains a map from
+// diective to its bitmask definitions, and a MatchFunc for it.
+//
+//go:embed tmpl/support_file.tmpl
+var supportFileTmplStr string
+
+//nolint:gochecknoglobals
+var supportFileTmpl = template.Must(template.New("supportFile").
+	Funcs(template.FuncMap{"Join": strings.Join}).Parse(supportFileTmplStr))
+
+//nolint:gochecknoglobals
+var ngxVarNameToGo = map[string]string{
+	"NGX_MAIL_MAIN_CONF":   "ngxMailMainConf",
+	"NGX_STREAM_MAIN_CONF": "ngxStreamMainConf",
+	"NGX_CONF_TAKE1":       "ngxConfTake1",
+	"NGX_STREAM_UPS_CONF":  "ngxStreamUpsConf",
+	"NGX_HTTP_LIF_CONF":    "ngxHTTPLifConf",
+	"NGX_CONF_TAKE2":       "ngxConfTake2",
+	"NGX_HTTP_UPS_CONF":    "ngxHTTPUpsConf",
+	"NGX_CONF_TAKE23":      "ngxConfTake23",
+	"NGX_CONF_TAKE12":      "ngxConfTake12",
+	"NGX_HTTP_MAIN_CONF":   "ngxHTTPMainConf",
+	"NGX_HTTP_LMT_CONF":    "ngxHTTPLmtConf",
+	"NGX_CONF_TAKE1234":    "ngxConfTake1234",
+	"NGX_MAIL_SRV_CONF":    "ngxMailSrvConf",
+	"NGX_CONF_FLAG":        "ngxConfFlag",
+	"NGX_HTTP_SRV_CONF":    "ngxHTTPSrvConf",
+	"NGX_CONF_1MORE":       "ngxConf1More",
+	"NGX_ANY_CONF":         "ngxAnyConf",
+	"NGX_CONF_TAKE123":     "ngxConfTake123",
+	"NGX_MAIN_CONF":        "ngxMainConf",
+	"NGX_CONF_NOARGS":      "ngxConfNoArgs",
+	"NGX_CONF_2MORE":       "ngxConf2More",
+	"NGX_CONF_TAKE3":       "ngxConfTake3",
+	"NGX_HTTP_SIF_CONF":    "ngxHTTPSifConf",
+	"NGX_EVENT_CONF":       "ngxEventConf",
+	"NGX_CONF_BLOCK":       "ngxConfBlock",
+	"NGX_HTTP_LOC_CONF":    "ngxHTTPLocConf",
+	"NGX_STREAM_SRV_CONF":  "ngxStreamSrvConf",
+	"NGX_DIRECT_CONF":      "ngxDirectConf",
+	"NGX_CONF_TAKE13":      "ngxConfTake13",
+	"NGX_CONF_ANY":         "ngxConfAny",
+	"NGX_CONF_TAKE4":       "ngxConfTake4",
+}
+
+//nolint:nonamedreturns
+func masksFromFile(path string) (directive2Masks map[string][]mask, err error) {
+	directive2Masks = make(map[string][]mask, 0)
+	byteContent, err := os.ReadFile(path)
```

Comment:
i'm wondering if we want to follow existing convention, by allowing reading of file from a FS (https://github.com/nginxinc/nginx-go-crossplane/blob/main/parse.go#L77), and by default open from the os. https://pkg.go.dev/os#DirFS. This would have the flexibility of using the embed for testing (https://pkg.go.dev/embed), instead of relying on reading the file system.

Reply:
Let's make sure first, do you mean test in this way?
```go
//go:embed testdata/*
var testFiles embed.FS
for _, test := range tests {
    readEmbed(test.path)
}
```
I thought currently user cases of the generator are to read from files and generate, and for unit tests I thought may not need to add files into compilation? Since these files are only used for one time in unit test, put them in compile won't be much faster then read using `os.Open`. I may caution adding to many unnecessary files(maybe some large file in future) into compilation. Maybe we can do that after the need of reading from sources other than local file system comes, or some ut reads files repeatedly and takes too much time?